File: proj8/proj_8_testers/tests/test4.ml
# Tests globals and scoping

begin
    # global needs to be defined before functions that use them!!!
    int x := 0;

    foo1() begin
        write("foo1: ", x, "\n");
        x := x + 10;
        write("foo1: ", x, "\n");
    end

    foo2() begin
        write("foo2: ", x, "\n");
    end

    foo3() begin
        int x := 300;
        write("foo3: ", x, "\n");
    end

    write("main: ", x, "\n");
    foo1()
    write("main: ", x, "\n");
    foo1()
    write("main: ", x, "\n");
    foo2()
    write("main: ", x, "\n");
    foo3()
    write("main: ", x, "\n");
end

File: archive/proj8/proj_8_testers/c_files/test4.c
// Test globals and scoping
#include <stdio.h>

int x;

void foo1() {
  printf("foo1: %d\n", x);
  x += 10;
  printf("foo1: %d\n", x);
}

void foo2() {
  printf("foo2: %d\n", x);
}

void foo3() {
  int x = 300;
  printf("foo3: %d\n", x);
}

int main() {
  x = 0;

  printf("main: %d\n", x);
  foo1();
  printf("main: %d\n", x);
  foo1();
  printf("main: %d\n", x);
  foo2();
  printf("main: %d\n", x);
  foo3();
  printf("main: %d\n", x);
  return 0;
}
